qcfractal/tests/test_services.py

```diff
@@ -120,6 +120,94 @@ def test_service_torsiondrive_duplicates(torsiondrive_fixture):
+    result = client.query_procedures(id=ret.ids)[0]
+    assert result.status == "COMPLETE"
+    # check final energies
+    final_energies = result.get_final_energies()
+    # the dihedral range should be limited to -150, -90, -60
+    assert set(final_energies.keys()) == {(-150,), (-120,), (-90,), (-60,)}
+    assert pytest.approx(0.0005683235570009067, abs=1e-6) == final_energies[(-150,)]
+    assert pytest.approx(0.0002170694130912583, abs=1e-6) == final_energies[(-120,)]
+    assert pytest.approx(0.0001565537585121726, abs=1e-6) == final_energies[(-90,)]
+    assert pytest.approx(0.0007991274441437338, abs=1e-6) == final_energies[(-60,)]
+    # check final molecules
+    final_molecules = result.get_final_molecules()
+    assert set(final_molecules.keys()) == {(-150,), (-120,), (-90,), (-60,)}
+    assert all(hasattr(m, "symbols") for m in final_molecules.values())
+
+    ## test optional energy_decrease_thresh feature
+    ret = spin_up_test(keywords={"grid_spacing": [90], "energy_decrease_thresh": 3e-5})
+    # Get a TorsionDriveORM result and check data
+    result = client.query_procedures(id=ret.ids)[0]
+    assert result.status == "COMPLETE"
+    # check final energies
+    final_energies = result.get_final_energies()
+    # the final energies are the same as the default setting, because this molecule is too simple
+    assert set(final_energies.keys()) == {(-90,), (-0,), (90,), (180,)}
+    assert pytest.approx(0.002597541340221565, abs=1e-6) == final_energies[(0,)]
+    assert pytest.approx(0.000156553761859276, abs=1e-6) == final_energies[(90,)]
+    assert pytest.approx(0.000156553761859271, abs=1e-6) == final_energies[(-90,)]
+    assert pytest.approx(0.000753492556057886, abs=1e-6) == final_energies[(180,)]
+    # check final molecules
+    final_molecules = result.get_final_molecules()
+    assert all(hasattr(m, "symbols") for m in final_molecules.values())
+
+    ## test optional energy_upper_limit feature
+    ret = spin_up_test(keywords={"grid_spacing": [30], "energy_upper_limit": 1e-4})
+    # Get a TorsionDriveORM result and check data
+    result = client.query_procedures(id=ret.ids)[0]
+    assert result.status == "COMPLETE"
+    # check final energies
+    final_energies = result.get_final_energies()
+    # the energy_upper_limit should limit the range of the scan
+    assert set(final_energies.keys()) == {(-150,), (-120,), (-90,), (-60,)}
+    assert pytest.approx(0.0005683235570009067, abs=1e-6) == final_energies[(-150,)]
+    assert pytest.approx(0.0002170694130912583, abs=1e-6) == final_energies[(-120,)]
+    assert pytest.approx(0.0001565537585121726, abs=1e-6) == final_energies[(-90,)]
+    assert pytest.approx(0.0007991274441437338, abs=1e-6) == final_energies[(-60,)]
+    # check final molecules
+    final_molecules = result.get_final_molecules()
+    assert all(hasattr(m, "symbols") for m in final_molecules.values())
+
+    ## test optional "extra_constraints" feature
```

Comment:
Could we try consolidating these tests? It should be sufficient to ensure that simply every option was touched rather than having a test that tries each optimization individually. As TD tests are some of our most expensive, keeping these down to a reasonable number is much preferable.

Reply:
I agree it's better to have short tests.  The additional tests added here takes about 6 seconds on my workstation. Another way to test here is to only parse these options without running actual jobs, however this will allow bugs in the implementation of these optional features to pass the test. Therefore I went the first and more robust way.

Reply:
Ok, sounds good. I suspect we hit lots of duplicate computations and the results are simply returned through here.